### ISSUE TITLE
[Feature request] Add support for Yubikey as auth backend

### ISSUE THREAD

**Issue by daker** [keeweb/keeweb#410]
It would be great if there was a way to be able to authenticate using Yubikey.

**Comment by antelle** [comment]
According to [this guide](http://keepass.info/help/kb/yubikey.html), Yubikey doesn't require any installation. Doesn't it work with KeeWeb as well?

**Comment by daker** [comment]
Yes KeeWeb works perfectly with the static password mode, but i guess adding support for one-time password will require more work, keepass has this feature as plugin.

**Comment by antelle** [comment]
Yes, one-time password will require more work. We may consider it in future...

**Comment by MAHDTech** [comment]
+1 would love to see OTP feature with Yubikey. It's the only thing stopping me using KeeWeb as I can't login to my databases when OTP is enabled.

**Comment by jmgilman** [comment]
Throwing in another +1, hardware OTP support is critical.

**Comment by martinl** [comment]
Hi, 

first thing to note is that OTP is meant to be used as second factor (in addition to password) authentication method for client-server applications and all methods are based on secret key shared between client (user) and server (trusted system). This shared secret should be stored securely externally (outside of KDBX file, encrypted or offline) for recovery purposes - a bit of a chicken and egg problem with password manager use case.

That aside, I did a quick review of Keepass OTP keyprovider plugins. Basically, there are two OTP options for using Keepass with Yubikey: HMAC-SHA1 challenge-response and OATH-HOTP.

HMAC-SHA1

HMAC-SHA1 challenge-response method is implemented by Keechallenge Keepass plugin (https://github.com/brush701/keechallenge/releases). Supporting Yubikey HMAC-SHA1 mode requires either Yubikey client library (https://developers.yubico.com/yubikey-personalization/) or low-level HID library with HID features support (for example https://github.com/node-hid/node-hid).

OATH-HOTP

OATH-HOTP is implemented by OtpKeyProv Keepass plugin (source: http://keepass.info/plugins.html#otpkeyprov). It is recommended to use at least four 8-digit OTP codes to create sufficiently strong key. 

For Keeweb integration OATH-HOTP likely the easiest OTP method to implement since OTP codes are sent as keyboard keypress events triggered by Yubikey button press. Also, HOTP has been implemented in pure javascript (https://github.com/guyht/notp) without any external dependencies.

Here is a guide for configuring Yubikey with Keepass OtpKeyProv:
https://www.yubico.com/why-yubico/for-individuals/password-managers/keepass/

OATH specs:
https://openauthentication.org/specifications-technical-resources/

I hope this small overview clarifies things a bit

**Comment by rmenessec** [comment]
As a user, HOTP is completely **terrible**. It's incompatible with having more than one YubiKey as a backup; if the YubiKey is destroyed / lost / stolen, HOTP will break instantly, and you'll lose access to the KeeWeb database forever.

Personally, I'm using HMAC-SHA1 with KeePassXC. This is (probably) the easiest way to implement YubiKey support from the perspective of both users and developers; the SHA1 secret can be copied to multiple YubiKeys, and the seed can be (for example) stored in a safety deposit box, or an encrypted filesystem, or whatever suits your security requirements. (The secret **CANNOT** be retrieved from the YubiKey directly.)

Basically, it's HMAC-SHA1 or #613.

... or there's Yubico's native OTP method, but that requires the use of Yubico's servers, for the most part, and Yubico OTP doesn't work with any other security devices that I'm aware of. It's possible to skip the use of Yubico's servers, but I'm reasonably sure that the alternative (hosting OTP authentication locally) requires buying at least one YubiHSM. They're not inexpensive. Well--they are compared to virtually every other HSM solution on the market, but they're not the kind of inexpensive where your typical home user is going to want to pay for them. Then there's the configuration, administration, _etc._

**Comment by a1exanderjung** [comment]
+1

**Comment by Babbleshack** [comment]
Any movement on this?

**Comment by 0xNeffarion** [comment]
+1 This would be great to have

**Comment by Ich-h4lt** [comment]
Any news of implementing hmac-sha1?

**Comment by BenoitDLX** [comment]
I'd love to see that feature too. I'm interested to use Keeweb but it won't work because i'm using the HMAC-SHA1 as 2FA.

**Comment by antelle** [comment]
Closing, see #1435, YubiKey support is coming soon
